src/ContrastAdjustmentAPI/intensity_adjustment.jl

```diff
@@ -0,0 +1,177 @@
+# usage example for package developer:
+#
+#     import ContrastAdjustmentAPI: AbstractIntensityAdjustmentAlgorithm,
+#                             adjust_intensity, adjust_intensity!
+
+"""
+    AbstractIntensityAdjustmentAlgorithm <: AbstractImageFilter
+
+A root type for `ImageContrastAdjustment` package that relates to algorithms
+that manipulate contrast without operating on intensity histograms.
+
+Any concrete intensity adjustment algorithm shall subtype it to support
+[`adjust_intensity`](@ref) and [`adjust_intensity!`](@ref) APIs.
+
+# Examples
+
+All intensity adjustment algorithms in ImageContrastAdjustment are called in the
+following pattern:
+
+```julia
+# first generate an algorithm instance
+f = LinearStretching()
+
+# then pass the algorithm to `adjust_intensity`
+img_adjusted = adjust_intensity(img, f)
+
+# or use in-place version `adjust_intensity!`
+img_adjusted = similar(img)
+adjust_intensity!(img_adjusted, img, f)
+```
+
+Some algorithms also receive additional information as an argument,
+e.g., `gamma` of `GammaCorrection`.
+
+```julia
+# you can explicit specify the parameters
+f = GammaCorrection(gamma = 1.4)
+```
+
+For more examples, please check [`adjust_intensity`](@ref),
+[`adjust_intensity!`](@ref) and concrete algorithms.
+"""
+abstract type AbstractIntensityAdjustmentAlgorithm <: AbstractImageFilter end

```

Comment:
Another choice is to make it 

```julia
abstract type AbstractIntensityAdjustmentAlgorithm <: AbstractHistogramAdjustmentAlgorithm end
```

and so that we don't need to deprecate the old usages. We can just incrementally add new `adjust_intensity` methods with default `alg` to be `LinearStretching`.

How do you think?

Reply:
I think that making

```julia
abstract type AbstractIntensityAdjustmentAlgorithm <: AbstractHistogramAdjustmentAlgorithm end
```

goes against the spirit of the issue https://github.com/JuliaImages/ImageContrastAdjustment.jl/issues/32 which seems to be specifically about the fact that operations that don't manipulate the histogram ought not to be conceptualised a histogram adjustment algorithms. 

I reckon let's just go ahead and make the deprecations before we embed this all in the Images ecosystem.

Reply:
> goes against the spirit of the issue #32 which seems to be specifically about the fact that operations that don't manipulate the histogram ought

My argument is that not all histogram adjustment algorithm needs to _explicitly_ construct a histogram. Directly adjust intensity can be considered as an _implicit_ way to scale the x scale of the intensity histogram.

My personal understanding of #32 is that we just need a default algorithm for easier and more convenient usage to close #32. For generic histogram adjustment, there isn't a consensus default algorithm, but for intensity adjustment, we can default to linear stretching.

Reply:
> My argument is that not all histogram adjustment algorithm needs to explicitly construct a histogram. Directly adjust intensity can be considered as an implicit way to scale the x scale of the intensity histogram.

That was also my argument and the reason why I implemented it the way I did in the first place :D, but @stillyslalom  seems to disagree with our view here. 

> We can just incrementally add new adjust_intensity methods with default alg to be LinearStretching. This means that both adjust_histogram and adjust_intensity would work for AbstractIntensityAdjustmentAlgorithm types.

I'm torn here. If we go this route, then what's the benefit of introducing the `adjust_intensity` function? Is it effectively not going to just be an alias for `adjust_histogram`? 

I guess one distinction is that `AbstractIntensityAdjustmentAlgorithm` is typically embarrassingly parallelisable, whereas some of the histogram-based algorithms may not be. 

> My personal understanding of #32 is that we just need a default algorithm for easier and more convenient usage to close #32. For generic histogram adjustment, there isn't a consensus default algorithm, but for intensity adjustment, we can default to linear stretching.

I was planning to introduce some convenience functions such as `span_contrast` to address issues such as https://github.com/JuliaImages/ImageContrastAdjustment.jl/issues/27

Reply:
I think all we need here is to introduce a convenient function `span_contrast`. Changing the type hierarchy doesn't make things easier.

Reply:
> My argument is that not all histogram adjustment algorithm needs to explicitly construct a histogram. Directly adjust intensity can be considered as an implicit way to scale the x scale of the intensity histogram.

I don't know if I agree with this argument. We don't call `fill!` `adjust_variance` even though it does that.

Reply:
Thanks for sharing your thoughts. What is your suggestion regarding the type hierarchy? Should AbstractIntensityAdjustmentAlgorithn subtype the AbstractHistogramAdjustmentAlgorithm type?

Reply:
Vice versa, I'd say. The second sounds way more specific, the first is much more vague.